Compose fixes for AWS machine

lower29 opened the pull request:
I've changed the `compose` build and configuration so that it works with `docker-machine` deploying to a remote docker host in AWS. This fixes #23, but also #15 and #18.

I've moved files around a lot, sorry. `shib` is gone, as there are no "abstract" images any more, just those in `am-shib` (nginx) and `shib-local` (idp and ldap). I've also deleted `example-shib` because it's not actually useful for what we're doing with Archivematica.

Aside from the reorganization, these changes overcome the following problems...

# Undeclared build dependencies
This I've overcome by using `docker run` to execute some tools and helper scripts within a docker container, thereby guaranteeing that the required dependencies are available. No more missing 'python' or 'readlink -f' commands.

# Dependency on files created earlier in the build process
It's still true that if you don't use `make` then you won't get a very satisfactory result. However, if you do use `make` now, then there are no dependencies on other parts of the build filesystem. The 'abstract' IdP build is gone, now integrated into the main 'am-shib' IdP.

The build process doesn't rely on as many scripts now either. Gone are the `build.sh`, and any `bootstrap.sh` are run within the container, through the `command` in the compose configuration. Also, rather than adding customized files to the image (requiring rebuild if they change), they are now added as volume mounts, further removing the need for extra scripting.

# Volume mounts don't work on a remote host
Fun fact: `docker-compose` resolves `.` locally during `docker-compose build` but remotely during `docker-compose up`. This means that `./am-shib/...` etc works perfectly during build but breaks horribly during deployment to a remote docker host. To fix this, all volume mount paths are now relative to `VOL_BASE`, which is set to `.` by default but should be set to the correct path on the remote host, e.g. `/home/ubuntu/rdss-archivematica/compose`.

# External ports for Dashboard and IdP need to be different
Previously Shib was configured to use 443 for both Dashboard and IdP, requiring them to be exposed on different IPs. This was fine locally, but in AWS we only have one interface. They now run on different ports by default (IdP is 6443 now), and are configurable via environment variables.

... I've also paved the way for using another IdP other than our own, by spliting the IdP into the `shib-local` compose config. This means we can choose not to deploy the `idp` and `ldap` containers, when the time comes to use UKAMF etc.

Last but not least, so-called "secrets", such as keys and certificates, are now done in `create-secrets` build targets and scripts (which will eventually become optional once proper keys/certs are available). This is a poor man's approach, but is a step in the right direction until we adopt Docker Secrets or Vault (or something else) at a later date. Currently passwords etc are still non-production-ready dummy values.

This has been tested against the AWS deployment instructions in `aws/machine`.

sevein commented on compose/README.md:
```diff
@@ -17,24 +17,38 @@ Service Sets
+	COMPOSE_FILE=docker-compose.dev.yml:docker-compose.am-shib.yml:docker-compose.shib-local.yml docker-compose \
+		<compose-args>
+
+This is quite a mouthful, so tere are also Makefiles defined that shortcut some of this for you, for example:
```
s/tere/here?

lower29 replied:
Well spotted, fixed.

lower29 commented on compose/am-shib/Makefile:
```diff
@@ -12,48 +14,42 @@ DOMAIN_NAME ?= "example.ac.uk"
+		--workdir "/src/nginx" \
+		rawmind/alpine-base:3.5-1 \
+		./create-secrets.sh
+
```
Using docker in this way ensures we have a known build environment. We specify the source and input directories using volume mounts, and do the same for the build output.

lower29 commented:
I've rebased against latest changes on master and pushed again.

lower29 commented:
I've rebased against latest master and pushed again. This is a forced push - you'll need to reset if you checked out previously.

lower29 commented on compose/docker-compose.dev.yml:
```diff
@@ -76,8 +76,9 @@ services:
-      - "./dev/etc/nginx/conf.d:/etc/nginx/conf.d:ro"
+      - "${VOL_BASE}/dev/etc/nginx/nginx.conf:/etc/nginx/nginx.conf:ro"
+      - "${VOL_BASE}/dev/etc/nginx/conf.d/archivematica.conf:/etc/nginx/conf.d/archivematica.conf:ro"
+      - "${VOL_BASE}/dev/etc/nginx/conf.d/default.conf:/etc/nginx/conf.d/default.conf:ro"
```
I had to basically revert the change from a previous commit that replaced specific files with the `conf.d` directory here. The reason for this is that we want to be able to remount different files for the `am-shib` `nginx` service, and you can't do that if you're mounting into a directory that is already mounted read-only.

Instead, I've put in an empty `default.conf`, which is what the original change was intending to replace anyway.

sevein replied:
That makes sense, sorry for the last minute change!

lower29 commented on compose/docker-compose.am-shib.yml:
```diff
@@ -74,27 +24,59 @@ services:
+      "--",
+      "/bin/sh", "-c",
+      "bootstrap.sh && /usr/bin/supervisord --nodaemon --configuration /etc/supervisor/supervisord.conf"
+    ]
```
Here we mount the `bootstrap.sh` script into the container and then modify the `command` so that it is called prior to the service daemon being started.

lower29 commented on compose/shib-local/etc/idp/jetty/backchannel.ini:
```diff
@@ -0,0 +1,8 @@
+--module=backchannel
+
+jetty.backchannel.port=8443
+jetty.backchannel.sslContext.keyStorePath=/secrets/idp-backchannel.p12
```
This new config file is necessary because we're setting a different `keyStorePath`, so that we can use our "secrets".

lower29 commented on compose/shib-local/ldap/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+# Copy the bootstrap script to a location where the Osixia container set up will
+# find it
+COPY bootstrap.sh /container/service/zz_bootstrap/finish.sh
```
The `osixia/openldap` image is based on `osixia/baseimage`, which has its own attempt at managing services within a container. Therefore, rather than running the bootstrap script and then starting the daemon, we install the bootstrap as an additional (short-lived) service within the container, utilising the Osixia infrastructure to run it.

This is also necessary because the intent of the bootstrap for LDAP - adding schema and users - needs to be done *after* the LDAP service has started, so we can't bootstrap and then start the daemon.

sevein commented:
When I build (`make all SHIBBOLETH_CONFIG=archivematica`) from a clean clone I see the error below. The workaround for me was to `mkdir -p compose/am-shib/build/idp`.

```
# Process each of the relevant compose dirs
make -C dev destroy ;  make -C am-shib destroy ;  make -C shib-local destroy ;
make[1]: Entering directory `/home/jesus/rdss-archivematica/compose/dev'
docker-compose stop
ERROR: build path /home/jesus/rdss-archivematica/compose/am-shib/build/idp either does not exist, is not accessible, or is not a valid URL.
make[1]: *** [destroy] Error 1
make[1]: Leaving directory `/home/jesus/rdss-archivematica/compose/dev'
make[1]: Entering directory `/home/jesus/rdss-archivematica/compose/am-shib'
# Stop containers and remove volumes
COMPOSE_FILE=/home/jesus/rdss-archivematica/compose/docker-compose.dev.yml:/home/jesus/rdss-archivematica/compose/docker-compose.am-shib.yml:/home/jesus/rdss-archivematica/compose/docker-compose.shib-local.yml docker-compose down --volumes
ERROR: build path /home/jesus/rdss-archivematica/compose/am-shib/build/idp either does not exist, is not accessible, or is not a valid URL.
make[1]: *** [destroy] Error 1
make[1]: Leaving directory `/home/jesus/rdss-archivematica/compose/am-shib'
make[1]: Entering directory `/home/jesus/rdss-archivematica/compose/shib-local'
# Stop containers and remove volumes
COMPOSE_FILE=/home/jesus/rdss-archivematica/compose/docker-compose.dev.yml:/home/jesus/rdss-archivematica/compose/docker-compose.am-shib.yml:/home/jesus/rdss-archivematica/compose/docker-compose.shib-local.yml docker-compose down --volumes
ERROR: build path /home/jesus/rdss-archivematica/compose/am-shib/build/idp either does not exist, is not accessible, or is not a valid URL.
make[1]: *** [destroy] Error 1
make[1]: Leaving directory `/home/jesus/rdss-archivematica/compose/shib-local'
make: *** [destroy] Error 2
```

lower29 commented:
The error is actually in this line in `docker-compose.dev.yml`:

```  idp:
    image: "arkivum/shibboleth-idp:${DOMAIN_NAME}"
    build:
      context: "./am-shib/build/idp/"```

The context should be `shib-local/build/idp` not `am-shib/build/idp`.

Good find! I obviously missed it when rearranging the code, sorry.

sevein commented:
Thanks, @lower29! I started again from scratch and Nginx is returning a 500 error and I'm seeing the following error in the terminal:

```
archivematica.example.ac.uk                     | 2017-06-18 21:49:36 DEBUG Shibboleth.Listener [3]: dispatching message (archivematica-dashboard/Login::run
::SAML2SI)
archivematica.example.ac.uk                     | 2017-06-18 21:49:36 WARN Shibboleth.SessionInitiator.SAML2 [3]: unable to locate metadata for provider (ht
tps://idp.example.ac.uk:6443/idp/shibboleth)
archivematica.example.ac.uk                     | 2017-06-18 21:49:36 WARN Shibboleth.SessionInitiator.SAML2 [3]: unable to locate metadata for provider (ht
tps://idp.example.ac.uk:6443/idp/shibboleth)
archivematica.example.ac.uk                     | 2017/06/18 21:49:36 [error] 908#0: *12 FastCGI sent in stderr: "shib:" while reading response header from
upstream, client: 172.18.0.1, server: archivematica.example.ac.uk, request: "GET / HTTP/1.1", subrequest: "/shibauthorizer", upstream: "fastcgi://unix:/tmp/
am-dashboard-shibauthorizer.sock:", host: "archivematica.example.ac.uk"
archivematica.example.ac.uk                     | 2017/06/18 21:49:36 [error] 908#0: *12 FastCGI sent in stderr: "Unable to locate metadata for identity pro
vider (https://idp.example.ac.uk:6443/idp/shibboleth)" while reading response header from upstream, client: 172.18.0.1, server: archivematica.example.ac.uk,
 request: "GET / HTTP/1.1", subrequest: "/shibauthorizer", upstream: "fastcgi://unix:/tmp/am-dashboard-shibauthorizer.sock:", host: "archivematica.example.a
c.uk"
archivematica.example.ac.uk                     | 172.18.0.1 - - [18/Jun/2017:21:49:36 +0000] "GET / HTTP/1.1" 500 192 "-" "Mozilla/5.0 (Macintosh; Intel Ma
c OS X 10.11; rv:53.0) Gecko/20100101 Firefox/53.0" "-"
```

ldap prints repeatedly the following:

```
ldap_1 | SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
ldap_1 | SASL SSF: 0
ldap_1 | 5946f5f4 conn=6432 op=1 ADD dn="cn=eduorg,cn=schema,cn=config"
ldap_1 | 5946f5f4 conn=6432 op=1 RESULT tag=105 err=80 text=olcAttributeTypes: Duplicate attributeType: "1.3.6.1.4.1.5923.1.2.1.2"
ldap_1 | ldap_add: Other (e.g., implementation specific) error (80)
ldap_1 |       additional info: olcAttributeTypes: Duplicate attributeType: "1.3.6.1.4.1.5923.1.2.1.2"
ldap_1 | 5946f5f4 conn=6432 op=2 UNBIND
ldap_1 | 5946f5f4 conn=6432 fd=16 closed
ldap_1 | adding new entry "cn=eduorg,cn=schema,cn=config"
ldap_1 |
ldap_1 | /container/run/process/zz_bootstrap
ldap_1 | 5946f5f4 conn=6433 fd=16 ACCEPT from IP=127.0.0.1:52866 (IP=0.0.0.0:389)
ldap_1 | 5946f5f4 conn=6433 op=0 BIND dn="cn=admin,dc=example,dc=ac,dc=uk" method=128
ldap_1 | 5946f5f4 conn=6433 op=0 BIND dn="cn=admin,dc=example,dc=ac,dc=uk" mech=SIMPLE ssf=0
ldap_1 | 5946f5f4 conn=6433 op=0 RESULT tag=97 err=0 text=
ldap_1 | 5946f5f4 conn=6433 op=1 ADD dn="ou=physics,dc=example,dc=ac,dc=uk"
ldap_1 | 5946f5f4 conn=6433 op=1 RESULT tag=105 err=68 text=
ldap_1 | ldap_add: Already exists (68)
ldap_1 | 5946f5f4 conn=6433 op=2 UNBIND
ldap_1 | 5946f5f4 conn=6433 fd=16 closed
ldap_1 | adding new entry "OU=physics,DC=example,DC=ac,DC=uk"
```

Am I doing something wrong?

sevein commented:
@mamedin, I've added you as a reviewer. I think it'd be nice to have a third person looking at this.

lower29 commented:
@sevein I've reproduced the nginx 500 error you describe. 😞

I'll investigate and do what I can to fix.

I've also reproduced the LDAP error; I have a fix for that already so will submit that as an update to this PR in due course.

Thanks for taking the time to review! :)

lower29 commented:
@sevein I believe I've fixed the nginx 500 problem and the ldap log problem in my latest commits.

The nginx error was annoyingly simple: the Makefile was specifying 443 as the default IdP port but everything else was using/expecting 6443. The net result was the IdP metadata (generated) was wrong, which is why the SP on the nginx node couldn't communicate with it.

The ldap error was due to the way the bootstrap script was being run. That container uses `runit`, which restarts "failed" services. Our bootstrap script was running once and then exiting, causing runit to restart it. The script didn't expect this so it was running again, without checking if it should update ldap before doing so - hence the errors. I've changed it so that it checks the state before executing the ldap commands, and it also now waits for SIGTERM before exiting, thereby avoiding a continuous loop.

sevein commented:
Forgot to push to this branch before pushing to master. But I've merged, thank you! It works!

The merge commit is 52da1d77f981e453f0d6f9780c4c94f734031b7b. Sorry, @lower29 I didn't want to squash your commits (wasn't really sure how to do it) but really wanted to merge this as it was much needed :)

lower29 commented:
Thanks @sevein! Don't worry about the commits, I just wanted to keep the history clean that's all.

I was going to use `git rebase -i HEAD~8` to squash, changing commit order so each of the fixups became part of the most "logical" commit from my original commits and the changing `pick` to `f`.

If we still want to tidy up in future then we can, albeit on `master`!
